src/KubernetesClient/Informers/SharedInformer.cs

```diff
@@ -0,0 +1,184 @@
+                                    .ToObservable()
+                                    .Concat(Observable.Never<ResourceEvent<TResource>>())
+                                    .ObserveOn(Scheduler.Immediate)
+                                    .Subscribe(observer);
+                            }
+
+                            if (type.HasFlag(ResourceStreamType.Watch))
+                                broadcasterAttachment = _masterObservable
+                                    // we could be ahead of broadcaster because we initialized from cache which gets updated before the message are sent to broadcaster
+                                    // this logic realigns us at the correct point with the broadcaster
+                                    .Do(x => _logger.LogTrace($"Received from broadcaster {x}"))
+                                    .SkipWhile(x => x.MessageNumber <= cacheSnapshot.Version)
+                                    .Select(x => x.Value)
+                                    .Do(x => _logger.LogTrace($"Aligned with broadcaster {x}"))
+                                    .SubscribeOn(_masterScheduler)
+                                    .ObserveOn(childScheduler)
+                                    .Subscribe(observer, () =>
+                                    {
+                                        _logger.LogTrace("Child OnComplete");
+                                        RemoveSubscriber();
+                                    });
+                            else
+                                observer.OnCompleted();
+
+                            // let broadcaster know we're done attaching to stream so it can resume it's regular work
+                            _logger.LogTrace("Finished attaching to stream - signalling to resume");
+                            lock (_lock)
+                            {
+                                _waitingSubscribers.Signal();
+                            }
+
+                            return broadcasterAttachment;
+                        })
+                        .ObserveOn(childScheduler)
+                        .SubscribeOn(childScheduler);
+                })
+                .SubscribeOn(childScheduler) // ensures that when we attach master observer it's done on child thread, as we plan on awaiting cache synchronization
+                .Do(_ => _logger.LogTrace($"Shared informer out: {_}"));
+        }
+
+
+        [DebuggerStepThrough]
+        private void YieldToWaitingSubscribers()
+        {
+            _logger.LogTrace("Waiting for subscribers to attach to stream");
+            while (_waitingSubscribers.CurrentCount > 0)
+                // give a chance to any joining subscribers to realign with the broadcast stream
+
+                _waitingSubscribers.Wait(100);
+            _logger.LogTrace("Finished yielding to subscribers");
+        }
+
+        private void AddSubscriber()
```

Comment:
The locking in `AddSubscriber` and `RemoveSubscriber` must be reconsidered given subscribers `SubscribeOn` and `ObserveOn` their own `EventLoopScheduler`, meaning the methods will be invoked concurrently. In particular, access to `_subscribers` must be serialised.